File: templating/core/src/main/python/jsonteng/tags/len_tag.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

from .tag_base import TagBase


class LenTag(TagBase):
    """
    Return the length of an object. The value depends on the object type.
    """
    name = "len"

    def __init__(self, tag_resolver):
        """
        Construct this tag.
        :param tag_resolver: Tag resolver
        :type tag_resolver: 'TagResolver'
        """
        super().__init__(tag_resolver)
        self._element_resolver = tag_resolver.get_element_resolver()
        self._template_loader = tag_resolver.get_template_loader()

    def process(self, tag_tokens, binding_data_list):
        """
        Process this tag.
        :param tag_tokens: Tag arguments.
        :type tag_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        if len(tag_tokens) != 1:
            raise Exception(
                "Tag \"{}\" requires 1 parameter. Parameters given {}".
                format(LenTag.name, tag_tokens))
        data = tag_tokens[0]
        resolved_data = self._element_resolver.resolve(data, binding_data_list)
        if resolved_data is not None:
            return len(resolved_data)
        else:
            return 0

File: templating/core/src/main/go/src/jsonteng/tags/len_tag.go
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package tags

import (
	"container/list"
	"fmt"
	"jsonteng/errors"
	intfs "jsonteng/interfaces"
)

type lenTag struct {
	intfs.TagBase
	elementResolver intfs.ElementResolver
}

func (tag *lenTag) init(tagResolver intfs.TagResolver) {
	tag.Name = "len"
	tag.TagResolver = tagResolver
	tag.elementResolver = tagResolver.GetElementResolver()
}

func (tag *lenTag) Process(tagTokens []interface{}, bindingDataList *list.List) (interface{}, error) {
	tokenCount := len(tagTokens)
	if tokenCount != 1 {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Tag \"%s\" requires 1 parameter. Parameters given %v", tag.Name, tagTokens)
		return nil, &errl
	}
	data := tagTokens[0]
	resolvedData, err := tag.elementResolver.ResolveElement(data, bindingDataList)
	if err != nil {
		return nil, err
	}
	return length(resolvedData), nil
}

func length(data interface{}) int {
	switch d := data.(type) {
	case map[string]interface{}:
		return len(d)
	case []interface{}:
		return len(d)
	default:
		return -1
	}
}
